src/dataflow/src/render/join/delta_join.rs

```diff
@@ -216,307 +212,261 @@ where
                 map_filter_project,
             );
 
-            for input in inputs.iter() {
```

Comment:
I suppose it didn't have to, but the convention for all operators is that the inputs are ensured rendered before the rendering is invoked. The join code already did this (lines 642ish in `render/mod.rs`) and I suspect this was vestigial. 

It allows each of the rendering fragments to be written with only their own type constraints, rather than the union of all constraints, is the philosophical reason for this.